File: app/routers/history.py
from __future__ import annotations

from fastapi import APIRouter, Request
from sqlalchemy import func, select

from app.deps import CurrentUser, CurrentWorkspace, DbSession
from app.models import ActivityLog, IndexJob, JobStatus, JobType, Site
from app.services.sites import list_sites
from app.services.stats import job_totals
from app.templating import render

router = APIRouter(prefix="/history", tags=["history"])

PER_PAGE = 60


@router.get("")
def history_page(
    request: Request,
    user: CurrentUser,
    workspace: CurrentWorkspace,
    db: DbSession,
    site_id: int = 0,
    status: str = "",
    job_type: str = "",
    page: int = 1,
):
    query = (
        select(IndexJob)
        .join(Site, Site.id == IndexJob.site_id)
        .where(Site.workspace_id == workspace.id)
    )
    count_query = (
        select(func.count(IndexJob.id))
        .join(Site, Site.id == IndexJob.site_id)
        .where(Site.workspace_id == workspace.id)
    )

    if site_id:
        query = query.where(IndexJob.site_id == site_id)
        count_query = count_query.where(IndexJob.site_id == site_id)
    if status:
        query = query.where(IndexJob.status == status)
        count_query = count_query.where(IndexJob.status == status)
    if job_type:
        query = query.where(IndexJob.job_type == job_type)
        count_query = count_query.where(IndexJob.job_type == job_type)

    total = db.scalar(count_query) or 0
    page = max(page, 1)
    jobs = list(
        db.scalars(
            query.order_by(IndexJob.created_at.desc())
            .offset((page - 1) * PER_PAGE)
            .limit(PER_PAGE)
        )
    )

    sites = list_sites(db, workspace)
    site_map = {site.id: site for site in sites}

    activity = list(
        db.scalars(
            select(ActivityLog)
            .where(ActivityLog.workspace_id == workspace.id)
            .order_by(ActivityLog.created_at.desc())
            .limit(25)
        )
    )

    return render(
        request,
        "history.html",
        {
            "user": user,
            "workspace": workspace,
            "jobs": jobs,
            "activity": activity,
            "sites": sites,
            "site_map": site_map,
            "total": total,
            "page": page,
            "pages": max(1, (total + PER_PAGE - 1) // PER_PAGE),
            "filter_site": site_id,
            "filter_status": status,
            "filter_type": job_type,
            "statuses": [s.value for s in JobStatus],
            "job_types": [t.value for t in JobType],
            "totals": job_totals(db, workspace.id, 30),
            "active_page": "history",
        },
    )

File: src/routes/history.ts
import { Router } from "express";
import { sql } from "kysely";
import { db } from "../db";
import { JobStatus, JobType } from "../db/types";
import { asyncHandler } from "../middleware/auth";
import { listSites } from "../services/sites";
import { jobTotals } from "../services/stats";
import { baseContext } from "../templating";
import { panelAuth } from "./auth";

export const historyRouter = Router();
historyRouter.use(...panelAuth);

const PER_PAGE = 60;

historyRouter.get(
  "/",
  asyncHandler(async (req, res) => {
    const workspace = req.workspace!;
    const siteId = Number(req.query.site_id) || 0;
    const status = typeof req.query.status === "string" ? req.query.status : "";
    const jobType = typeof req.query.job_type === "string" ? req.query.job_type : "";
    const page = Math.max(1, Number(req.query.page) || 1);

    let query = db
      .selectFrom("index_jobs")
      .innerJoin("sites", "sites.id", "index_jobs.site_id")
      .where("sites.workspace_id", "=", workspace.id);
    if (siteId) query = query.where("index_jobs.site_id", "=", siteId);
    if (status) query = query.where("index_jobs.status", "=", status as JobStatus);
    if (jobType) query = query.where("index_jobs.job_type", "=", jobType as JobType);

    const totalRow = await query
      .select(sql<number>`COUNT(index_jobs.id)`.as("total"))
      .executeTakeFirst();
    const total = Number(totalRow?.total ?? 0);

    const jobs = await query
      .selectAll("index_jobs")
      .orderBy("index_jobs.created_at", "desc")
      .offset((page - 1) * PER_PAGE)
      .limit(PER_PAGE)
      .execute();

    const sites = await listSites(workspace.id);
    const siteMap = Object.fromEntries(sites.map((s) => [s.id, s]));

    const activity = await db
      .selectFrom("activity_log")
      .selectAll()
      .where("workspace_id", "=", workspace.id)
      .orderBy("created_at", "desc")
      .limit(25)
      .execute();

    res.render(
      "history.html",
      baseContext(req, {
        user: req.user,
        workspace,
        jobs,
        activity,
        sites,
        site_map: siteMap,
        total,
        page,
        pages: Math.max(1, Math.ceil(total / PER_PAGE)),
        filter_site: siteId,
        filter_status: status,
        filter_type: jobType,
        statuses: Object.values(JobStatus),
        job_types: Object.values(JobType),
        totals: await jobTotals(workspace.id, 30),
        active_page: "history",
      }),
    );
  }),
);
